File: 2022/01/ruby/solution.rb
#!/usr/bin/env ruby
text = File.read("input.txt")

# list of list of integer
data = text.split("\n\n").map do |elf|
  elf.split("\n").map(&:to_i)
end

totals = data.map(&:sum)
max = totals.max
puts "Max calories: #{max}"

sorted = totals.sort
top = sorted[-3..-1] # bug: totals[-3..-1]
top_sum = top.sum
puts "Sum of top 3 elves calories: #{top_sum}"

File: 2022/01/python/solution.py
#!/usr/bin/env python3
import sys

class Calories(object):
    def __init__(self, input):
        elves = input.split("\n\n")
        self.elves = [sum(map(int, e.split())) for e in elves]
        # print(self.elves)

    def max(self):
        return max(self.elves)

    def top3sum(self):
        sorted = self.elves[:]
        sorted.sort()           # bug: expecting a return value (Ruby)
        top3 = sorted[-3:]      # bug: [-3:-1]  (Ruby)
        return sum(top3)

if __name__ == '__main__':
    fn = sys.argv[1] if len(sys.argv) > 1 else "input.txt"
    with open(fn) as f:
        input = f.read()
    sol = Calories(input)

    print("Part 1, max calories")
    print(sol.max())

    print("Part 2, sum of top 3 calories")
    print(sol.top3sum())
